modules/nextflow/src/main/groovy/nextflow/util/ArrayBag.groovy

```diff
@@ -63,6 +63,16 @@ class ArrayBag<E> implements Bag<E>, List<E>, KryoSerializable {
         InvokerHelper.inspect(this)
     }
 
+   @Override
+   int hashCode() {
+       target.hashCode()
+   }
+
+   @Override
+   boolean equals(Object o) {
+       target.equals(o)
+   }
```

Comment:
There's a reason why these methods were commented out, related to some Groovy bug, but now I can't recall what it was

Reply:
I suspected that might have been the case (reasonable looking code that is commented out is always a bit ominous).
The changes don't break any tests, if that gives you any confidence.

Reply:
Still not sure to understand \*how* it solves the problem

Reply:
Yeah, me neither. I'll carve out some time from customer things this week to try and dig into the *why*.

Reply:
This is the Groovy issue I was thinking of: https://github.com/nextflow-io/nextflow/issues/5187#issuecomment-2262906488

@pditommaso maybe it would be good to convert ArrayBag to Java, so that we don't have to deal with this weird Groovy behavior

Reply:
it's not that weird, the semantic of equals and hashCode in groovy is based on the collection content, not the object id, as it's in Java. 

It should be enough to implement the same logic as it's has been done [here](https://github.com/nextflow-io/nextflow/blob/71f6ed0bcda3462c415a23fb92c14f020ac55bce/modules/nf-commons/src/main/nextflow/util/HashBuilder.java#L159-L159), to make it work. 


@robsyme can you please provide a unit test to validate your solution?

Reply:
I mean a unit test for JoinOp, not the ArrayBag class 

https://github.com/nextflow-io/nextflow/blob/5c9e305d8a506c980d2fb086b6c90f482ecd12fb/modules/nextflow/src/test/groovy/nextflow/extension/JoinOpTest.groovy#L29-L29

Reply:
Added the JoinOpTest unit test, thanks Paolo.